src/lucas_kanade.jl

```diff
@@ -1,259 +1,240 @@
-    first_axis, second_axis = axes(img)
-    allowed_area = map(i -> first(i) + window_size:last(i) - window_size, axes(img))
-    new_point = point + displacement
-
-    if !lies_in(allowed_area, new_point) || !lies_in(allowed_area, point)
-        is_truncated_window = true
-    else
-        is_truncated_window = false
-    end
+function compute_spatial_gradient(pyramid::LKPyramid, grid, level)
+    G = _compute_spatial_gradient(
+        grid, pyramid.Iyy[level], pyramid.Iyx[level], pyramid.Ixx[level],
+    )
+    U, S, V = G |> svd2x2
+    G_inv = pinv2x2(U, S, V)
+    min_eigenvalue = min(S[1, 1], S[2, 2]) / prod(length.(grid))
 
-    # TODO: Convert these loops to explicit formulae.
-    w_up = 0
-    while point[1] - (w_up + 1) > first(first_axis)  && new_point[1] - (w_up + 1) >  first(first_axis)  &&  (w_up + 1) < window_size
-        w_up += 1
-    end
+    G_inv, min_eigenvalue
+end
 
-    w_down = 0
-    while point[1] + (w_down + 1)  < last(first_axis)  && new_point[1] + (w_down + 1) <  last(first_axis)  && (w_down + 1) < window_size
-        w_down += 1
-    end
+function prepare_linear_system(corresponding_point, A, Iy, Ix, offsets, B)
+    P, Q = size(A)
+    by, bx = 0.0, 0.0
 
-    w_left = 0
-    while point[2] - (w_left + 1) > first(second_axis) && new_point[2] - (w_left + 1) >  first(second_axis) && (w_left + 1) < window_size
-        w_left += 1
-    end
+    @inbounds for q in 1:Q, p in 1:P
+        r = corresponding_point[1] + offsets[1][p]
+        c = corresponding_point[2] + offsets[2][q]
 
-    w_right = 0
-    while point[2] + (w_right + 1) < last(second_axis) && new_point[2] + (w_right + 1) >  last(second_axis) && (w_right + 1) < window_size
-        w_right += 1
+        δI = A[p, q] - B(r, c)
+        by += δI * Iy[p, q]
+        bx += δI * Ix[p, q]
     end
 
+    SVector{2, Float64}(by, bx)
+end
 
-    grid_1 = (point[1] - w_up:point[1] + w_down, point[2] - w_left:point[2] + w_right)
-    offsets = (UnitRange(-w_up, w_down), UnitRange(-w_left, w_right))
+function compute_flow_vector(
+    corresponding_point,
+    first_pyramid::LKPyramid, etp, level,
+    grid, offsets, G_inv,
+)
+    b = prepare_linear_system(
+        corresponding_point,
+        view(first_pyramid.layers[level], grid[1], grid[2]),
+        view(first_pyramid.Iy[level], grid[1], grid[2]),
+        view(first_pyramid.Ix[level], grid[1], grid[2]),
+        offsets, etp,
+    )
+    G_inv * b
+end
 
-    return grid_1, offsets, is_truncated_window
+@inline is_lost(displacement, window) =
+    @inbounds(displacement[1] > window || displacement[2] > window)
+
+@inline lies_in(area, point) = @inbounds(
+    first(area[1]) ≤ point[1] ≤ last(area[1]) &&
+    first(area[2]) ≤ point[2] ≤ last(area[2])
+)
+
+"""
+# Arguments
+- `level`: Level of the pyramid in `[1, levels]` range.
+"""
+@inline get_pyramid_coordinate(point, level) = floor.(Int64, point ./ 2 ^ (level - 1))
```

Comment:
Unless there's a real reason, let's use `Int` here:

```suggestion
@inline get_pyramid_coordinate(point, level) = floor.(Int, point ./ 2 ^ (level - 1))
```